samples/ClientSample/HubSample.cs

```diff
@@ -48,71 +48,102 @@ internal static void Register(CommandLineApplication app)
 
             var connection = connectionBuilder.Build();
 
-            try
+            Console.CancelKeyPress += (sender, a) =>
             {
-                var closeTcs = new TaskCompletionSource<object>();
-                connection.Closed += e => closeTcs.SetResult(null);
-                // Set up handler
-                connection.On<string>("Send", Console.WriteLine);
+                a.Cancel = true;
+                connection.DisposeAsync().GetAwaiter().GetResult();
+            };
 
-                await ConnectAsync(connection);
+            // Set up handler
+            connection.On<string>("Send", Console.WriteLine);
 
-                Console.WriteLine("Connected to {0}", uri);
+            CancellationTokenSource closedTokenSource = null;
 
-                var sendCts = new CancellationTokenSource();
+            connection.Closed += e =>
+            {
+                // This should never be null by the time this fires
+                closedTokenSource.Cancel();
+
+                Console.WriteLine("Connection closed...");
+            };
+
+            while (true)
+            {
+                // Dispose the previous token
+                closedTokenSource?.Dispose();
+
+                // Create a new token for this run
+                closedTokenSource = new CancellationTokenSource();
 
-                Console.CancelKeyPress += async (sender, a) =>
+                // Connect to the server
+                if (!await ConnectAsync(connection))
                 {
-                    a.Cancel = true;
-                    Console.WriteLine("Stopping loops...");
-                    sendCts.Cancel();
-                    await connection.DisposeAsync();
-                };
+                    break;
+                }
+
+                Console.WriteLine("Connected to {0}", uri); ;
 
-                while (!closeTcs.Task.IsCompleted)
+                // Handle the connected connection
+                while (true)
                 {
-                    var completedTask = await Task.WhenAny(Task.Run(() => Console.ReadLine()), closeTcs.Task);
-                    if (completedTask == closeTcs.Task)
+                    try
                     {
-                        break;
-                    }
+                        var line = Console.ReadLine();
 
-                    var line = await (Task<string>)completedTask;
+                        if (line == null || closedTokenSource.Token.IsCancellationRequested)
+                        {
+                            break;
+                        }
 
-                    if (line == null)
+                        await connection.InvokeAsync<object>("Send", line);
+                    }
+                    catch (IOException)
```

Comment:
I don't like that we have to catch various kinds of exceptions to detect restartable conditions... Not sure the best way to approach that though.